quantum/backlight/backlight_arm.c

```diff
@@ -0,0 +1,266 @@
+    #define USE_GPIOV1
+#endif
+
+//GPIOV2 && GPIOV3
+#ifndef BACKLIGHT_PAL_MODE
+    #define BACKLIGHT_PAL_MODE      2
+#endif
+
+// GENERIC
+#ifndef BACKLIGHT_PWM_DRIVER
+    #define BACKLIGHT_PWM_DRIVER    PWMD3
+#endif
+#ifndef BACKLIGHT_PWM_CHANNEL
+    #define BACKLIGHT_PWM_CHANNEL    3
+#endif
+
+static void breathing_callback(PWMDriver *pwmp);
+
+static PWMConfig pwmCFG = {
+  0xFFFF,                           /* PWM clock frequency  */
+  256,                              /* PWM period (in ticks) 1S (1/10kHz=0.1mS 0.1ms*10000 ticks=1S) */
+  NULL,                             /* No Callback */
+  {                                 /* Default all channels to disabled - Channels will be configured durring init */
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL}
+  },
+  0,                                /* HW dependent part.*/
+  0
+};
+
+static PWMConfig pwmCFG_breathing = {
+  0xFFFF,                           /** PWM clock frequency  */
+  256,                              /* PWM period (in ticks) 1S (1/10kHz=0.1mS 0.1ms*10000 ticks=1S) */
+  breathing_callback,               /* Breathing Callback */
+  {                                 /* Default all channels to disabled - Channels will be configured durring init */
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL},
+      {PWM_OUTPUT_DISABLED, NULL}
+  },
+  0,                                /* HW dependent part.*/
+  0
+};
+
+// See http://jared.geek.nz/2013/feb/linear-led-pwm
+static uint16_t cie_lightness(uint16_t v) {
+  if (v <= 5243) // if below 8% of max
+    return v / 9; // same as dividing by 900%
+  else {
+    uint32_t y = (((uint32_t) v + 10486) << 8) / (10486 + 0xFFFFUL); // add 16% of max and compare
+    // to get a useful result with integer division, we shift left in the expression above
+    // and revert what we've done again after squaring.
+    y = y * y * y >> 8;
+    if (y > 0xFFFFUL) // prevent overflow
+      return 0xFFFFU;
+    else
+      return (uint16_t) y;
+  }
+}
+
+void backlight_init_ports(void) {
+  //printf("backlight_init_ports()\n");
+
+#ifdef USE_GPIOV1
+  palSetPadMode(PAL_PORT(BACKLIGHT_PIN), PAL_PAD(BACKLIGHT_PIN), PAL_MODE_STM32_ALTERNATE_PUSHPULL);
+#else
+  palSetPadMode(PAL_PORT(BACKLIGHT_PIN), PAL_PAD(BACKLIGHT_PIN), PAL_MODE_ALTERNATE(BACKLIGHT_PAL_MODE));
+#endif
+
+  pwmCFG.channels[BACKLIGHT_PWM_CHANNEL - 1].mode = PWM_OUTPUT_ACTIVE_HIGH;
+  pwmCFG_breathing.channels[BACKLIGHT_PWM_CHANNEL - 1].mode = PWM_OUTPUT_ACTIVE_HIGH;
+  pwmStart(&BACKLIGHT_PWM_DRIVER, &pwmCFG);
+
+  backlight_set(get_backlight_level());
+  if (is_backlight_breathing()) {
+    breathing_enable();
+  }
+}
+
+void backlight_set(uint8_t level) {
+    //printf("backlight_set(%d)\n", level);
+    uint32_t duty = (uint32_t)(cie_lightness(0xFFFF * (uint32_t) level / BACKLIGHT_LEVELS));
```

Comment:
maybe move this inside the if clause, as it only affects 1 branch, to avoid calc when backlight is off

Reply:
Moved